juju/juju
# [JUJU-3306] Ensure that a target controller has the expected external controller records created

If a source controller hosts consuming and offering models with a cross model relation, then migrating the consuming model to a new controller breaks because the model on the target controller now has relation(s) to an external controller (it previously was the same controller). But the migration process does not create the necessary external controller record.

Here we extend what the final model activate api call does - it is given the details of the source controller and any hosted models on that controller to which the migrated model has relations. This info is used to create the external controller record. We also need to ensure that any relevant remote application records has the source controller correctly set for reference counting purposes.

The migration target facade gets a version bump. We need to add a precheck so that if the target controller does not have the version needed, and there is an external controller record that would need to be created, the migration fails.

Also removed unused code.

## Checklist


## QA steps

bootstrap
add-model o
deploy juju-qa-dummy-source
offer dummy-source:sink
add-model c
deploy juju-qa-dummy-sink
relate dummy-sink o.dummy-source

bootstrap a new controller
migrate model c to new controller
check status

check logs on target controller to ensure no errors - previously the remote application worker would error
check the db to see that an external controller record is created and that the remote app has the source controller uuid set

## Bug reference

https://bugs.launchpad.net/bugs/2011662
